dojo/endpoint/utils.py

```diff
@@ -0,0 +1,255 @@
+    if kwargs.get('protocol'):
+        qs = qs.filter(protocol__iexact=kwargs['protocol'])
+    else:
+        qs = qs.filter(protocol__isnull=True)
+
+    if kwargs.get('userinfo'):
+        qs = qs.filter(userinfo__exact=kwargs['userinfo'])
+    else:
+        qs = qs.filter(userinfo__isnull=True)
+
+    if kwargs.get('host'):
+        qs = qs.filter(host__iexact=kwargs['host'])
+    else:
+        qs = qs.filter(host__isnull=True)
+
+    if kwargs.get('port'):
+        if (kwargs.get('protocol')) and \
+                (kwargs['protocol'].lower() in SCHEME_PORT_MAP) and \
+                (SCHEME_PORT_MAP[kwargs['protocol'].lower()] == kwargs['port']):
+            qs = qs.filter(Q(port__isnull=True) | Q(port__exact=SCHEME_PORT_MAP[kwargs['protocol'].lower()]))
+        else:
+            qs = qs.filter(port__exact=kwargs['port'])
+    else:
+        if (kwargs.get('protocol')) and (kwargs['protocol'].lower() in SCHEME_PORT_MAP):
+            qs = qs.filter(Q(port__isnull=True) | Q(port__exact=SCHEME_PORT_MAP[kwargs['protocol'].lower()]))
+        else:
+            qs = qs.filter(port__isnull=True)
+
+    if kwargs.get('path'):
+        qs = qs.filter(path__exact=kwargs['path'])
+    else:
+        qs = qs.filter(path__isnull=True)
+
+    if kwargs.get('query'):
+        qs = qs.filter(query__exact=kwargs['query'])
+    else:
+        qs = qs.filter(query__isnull=True)
+
+    if kwargs.get('fragment'):
+        qs = qs.filter(fragment__exact=kwargs['fragment'])
+    else:
+        qs = qs.filter(fragment__isnull=True)
+
+    if kwargs.get('product'):
+        qs = qs.filter(product__exact=kwargs['product'])
+    elif kwargs.get('product_id'):
+        qs = qs.filter(product_id__exact=kwargs['product_id'])
+    else:
+        qs = qs.filter(product__isnull=True)
+
+    return qs
+
+
+def endpoint_get_or_create(**kwargs):
+
+    qs = endpoint_filter(**kwargs)
+
+    if qs.count() == 0:
+        return Endpoint.objects.get_or_create(**kwargs)
+
+    elif qs.count() == 1:
+        return qs.first(), False
+
+    else:
+        raise MultipleObjectsReturned()
+
+
+def clean_hosts_run(apps, change):
+    def err_log(message, html_log, endpoint_html_log, endpoint):
+        error_suffix = 'It is not possible to migrate it. Delete or edit this endpoint.'
+        html_log.append({**endpoint_html_log, **{'message': message}})
+        logger.error('Endpoint (id={}) {}. {}'.format(endpoint.pk, message, error_suffix))
+        broken_endpoints.add(endpoint.pk)
+    html_log = []
+    broken_endpoints = set()
+    Endpoint_model = apps.get_model('dojo', 'Endpoint')
+    Endpoint_Status_model = apps.get_model('dojo', 'Endpoint_Status')
+    Product_model = apps.get_model('dojo', 'Product')
+    # when run on older versions, we haven't had the migration applied which has userinfo field added
+    # so defer that field and order by id (default ordering references userinfo field)
+    for endpoint in Endpoint_model.objects.defer('userinfo').order_by('id'):
```

Comment:
the defer userinfo is no longer needed if we are not backporting the script to 1.15.x

Reply:
done